File: public/localvault-pc-server.js
import express from 'express';
import multer from 'multer';
import fs from 'fs';
import path from 'path';
import crypto from 'crypto';
import os from 'os';

/**
 * LocalVault - PC-Side External SSD Photo Backup Server
 * 
 * Run on your Windows PC (or Mac/Linux) where your external SSD is connected.
 * 
 * Prerequisites:
 *   1. Node.js installed (https://nodejs.org)
 *   2. Run: npm install express multer
 *   3. Connect your External SSD (e.g. D:\ or E:\ on Windows)
 *   4. Run: node localvault-pc-server.js
 */

const PORT = 3000;

// Set your External SSD folder path below:
// Windows example: 'D:\\iPhone_Photos_SSD' or 'E:\\Photos_Vault'
// Mac example: '/Volumes/Samsung_T7/Photos_Vault'
// Linux example: '/media/user/SSD/Photos_Vault'
const SSD_MOUNT_PATH = process.env.SSD_PATH || (process.platform === 'win32' ? 'D:\\iPhone_Photos_SSD' : path.join(os.homedir(), 'iPhone_Photos_SSD'));

console.log('====================================================');
console.log('  LocalVault - Dual iPhone Auto Wi-Fi SSD Server');
console.log('====================================================');
console.log(`Target SSD Folder: ${SSD_MOUNT_PATH}`);

// Ensure SSD folder exists
try {
  fs.mkdirSync(SSD_MOUNT_PATH, { recursive: true });
  console.log(`[OK] SSD target directory confirmed.`);
} catch (err) {
  console.warn(`[WARN] Could not create folder at ${SSD_MOUNT_PATH}. Make sure your SSD is plugged in!`, err.message);
}

// Keep a local manifest of SHA-256 hashes to prevent duplicate uploads
const MANIFEST_FILE = path.join(SSD_MOUNT_PATH, '.localvault_manifest.json');
let knownHashes = new Set();

try {
  if (fs.existsSync(MANIFEST_FILE)) {
    const raw = fs.readFileSync(MANIFEST_FILE, 'utf-8');
    const list = JSON.parse(raw);
    knownHashes = new Set(list);
    console.log(`[Index] Loaded ${knownHashes.size} known photo hashes for deduplication.`);
  }
} catch (e) {
  knownHashes = new Set();
}

function saveManifest() {
  try {
    fs.writeFileSync(MANIFEST_FILE, JSON.stringify(Array.from(knownHashes), null, 2));
  } catch (err) {
    // ignore
  }
}

// Find PC Local IP on your Wi-Fi
function getLanIp() {
  const nets = os.networkInterfaces();
  for (const name of Object.keys(nets)) {
    for (const net of nets[name] || []) {
      if (net.family === 'IPv4' && !net.internal) {
        return net.address;
      }
    }
  }
  return 'localhost';
}

const LAN_IP = getLanIp();

const app = express();
app.use(express.json({ limit: '50mb' }));

// Multer storage engine - writes directly to SSD
const storage = multer.diskStorage({
  destination: (req, file, cb) => {
    const deviceName = (req.body.deviceName || req.body.deviceId || 'iPhone').replace(/[^a-zA-Z0-9_-]/g, '_');
    const takenDateStr = req.body.takenAt || new Date().toISOString();
    const d = new Date(takenDateStr);
    const year = isNaN(d.getTime()) ? '2026' : d.getFullYear().toString();
    const month = isNaN(d.getTime()) ? '09' : String(d.getMonth() + 1).padStart(2, '0');

    const destDir = path.join(SSD_MOUNT_PATH, deviceName, year, month);
    fs.mkdirSync(destDir, { recursive: true });
    cb(null, destDir);
  },
  filename: (req, file, cb) => {
    const timestamp = new Date().toISOString().replace(/[:.]/g, '-').slice(0, 19);
    const safeName = file.originalname.replace(/[^a-zA-Z0-9_.-]/g, '_');
    cb(null, `${timestamp}_${safeName}`);
  }
});

const upload = multer({ storage });

// Health check endpoint for iPhone
app.get('/api/health', (req, res) => {
  res.json({
    status: 'online',
    server: 'LocalVault PC SSD Server',
    ssdPath: SSD_MOUNT_PATH,
    lanIp: LAN_IP,
    totalPhotosIndexed: knownHashes.size
  });
});

// Deduplication check endpoint
app.post('/api/backup/check', (req, res) => {
  const hashes = req.body.hashes || [];
  const neededHashes = [];
  const duplicateHashes = [];

  for (const h of hashes) {
    if (knownHashes.has(h)) {
      duplicateHashes.push(h);
    } else {
      neededHashes.push(h);
    }
  }

  res.json({
    neededHashes,
    duplicateHashes,
    skippedDuplicatesCount: duplicateHashes.length
  });
});

// Upload endpoint
app.post('/api/backup/upload', upload.single('file'), (req, res) => {
  const file = req.file;
  if (!file) {
    return res.status(400).json({ error: 'No file received' });
  }

  // Calculate file hash
  let fileHash = '';
  try {
    const buf = fs.readFileSync(file.path);
    fileHash = crypto.createHash('sha256').update(buf).digest('hex');
  } catch {
    fileHash = crypto.createHash('sha256').update(file.originalname + file.size).digest('hex');
  }

  // Skip if duplicate
  if (knownHashes.has(fileHash)) {
    try { fs.unlinkSync(file.path); } catch (e) {}
    return res.json({
      status: 'duplicate_skipped',
      message: 'Photo already exists on SSD'
    });
  }

  knownHashes.add(fileHash);
  saveManifest();

  console.log(`[Backup Saved] -> ${file.path} (${(file.size / (1024 * 1024)).toFixed(2)} MB)`);

  res.json({
    status: 'synced',
    filename: file.filename,
    path: file.path,
    size: file.size
  });
});

app.listen(PORT, '0.0.0.0', () => {
  console.log('----------------------------------------------------');
  console.log(`Server is LIVE on your Home Wi-Fi!`);
  console.log(`From your 2 iPhones, access: http://${LAN_IP}:${PORT}`);
  console.log(`Direct SSD upload endpoint:  http://${LAN_IP}:${PORT}/api/backup/upload`);
  console.log('----------------------------------------------------');
});

File: public/localvault-pc-server.py
import os
import sys
import hashlib
from datetime import datetime
from pathlib import Path
from flask import Flask, request, jsonify

# LocalVault - Python Server for PC with Connected SSD
# 
# How to run on your PC:
#   1. Install python (if not installed)
#   2. Run: pip install flask
#   3. Run: python localvault-pc-server.py
#
# No Raspberry Pi needed! Runs directly on your Windows, Mac, or Linux PC.

PORT = 3000

# Set your external SSD path here:
# Windows: r"D:\iPhone_Photos_SSD" or r"E:\iPhone_Photos_SSD"
# Mac: "/Volumes/Samsung_T7/iPhone_Photos_SSD"
SSD_PATH = Path(r"D:\iPhone_Photos_SSD") if sys.platform == "win32" else Path.home() / "iPhone_Photos_SSD"
SSD_PATH.mkdir(parents=True, exist_ok=True)

MANIFEST_FILE = SSD_PATH / ".localvault_manifest.txt"
known_hashes = set()

if MANIFEST_FILE.exists():
    with open(MANIFEST_FILE, "r") as f:
        known_hashes = set(line.strip() for line in f if line.strip())

app = Flask(__name__)

@app.route('/api/health', methods=['GET'])
def health():
    return jsonify({
        "status": "online",
        "ssdPath": str(SSD_PATH),
        "totalPhotos": len(known_hashes)
    })

@app.route('/api/backup/upload', methods=['POST'])
def upload_file():
    if 'file' not in request.files:
        return jsonify({"error": "No file uploaded"}), 400
    
    file = request.files['file']
    device_name = request.form.get('deviceName', request.form.get('deviceId', 'iPhone'))
    safe_device = "".join(c for c in device_name if c.isalnum() or c in ('_', '-'))
    
    now = datetime.now()
    year = now.strftime("%Y")
    month = now.strftime("%m")
    
    target_dir = SSD_PATH / safe_device / year / month
    target_dir.mkdir(parents=True, exist_ok=True)
    
    timestamp = now.strftime("%Y%m%d_%H%M%S")
    safe_filename = f"{timestamp}_{file.filename}"
    target_filepath = target_dir / safe_filename
    
    file.save(str(target_filepath))
    
    # Calculate SHA-256
    hasher = hashlib.sha256()
    with open(target_filepath, 'rb') as f:
        while chunk := f.read(65536):
            hasher.update(chunk)
    file_hash = hasher.hexdigest()
    
    # Check deduplication
    if file_hash in known_hashes:
        target_filepath.unlink(missing_ok=True)
        return jsonify({"status": "duplicate_skipped", "message": "Photo already on SSD"})
    
    known_hashes.add(file_hash)
    with open(MANIFEST_FILE, "a") as f:
        f.write(file_hash + "\n")
        
    print(f"Backed up to SSD: {target_filepath}")
    return jsonify({
        "status": "synced",
        "filename": safe_filename,
        "path": str(target_filepath)
    })

if __name__ == '__main__':
    print(f"LocalVault Python PC Server running! SSD Path: {SSD_PATH}")
    print(f"Port: {PORT}")
    app.run(host='0.0.0.0', port=PORT)
